libs/x-content/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentParser.java

```diff
@@ -166,6 +167,26 @@ public long doLongValue() throws IOException {
         return parser.getLongValue();
     }
 
+    @Override
+    protected long doUnsignedLongValue() throws IOException {
+        JsonParser.NumberType numberType = parser.getNumberType();
+        if ((numberType == JsonParser.NumberType.INT) || (numberType == JsonParser.NumberType.LONG)) {
+            long longValue = parser.getLongValue();
+            if (longValue < 0) {
+                throw new IllegalArgumentException("Value [" + longValue + "] is out of range for unsigned long.");
+            }
+            return longValue;
+        } else if (numberType == JsonParser.NumberType.BIG_INTEGER) {
+            BigInteger bigIntegerValue = parser.getBigIntegerValue();
+            if (bigIntegerValue.compareTo(BIGINTEGER_MAX_UNSIGNED_LONG_VALUE) > 0 || bigIntegerValue.compareTo(BigInteger.ZERO) < 0) {
+                throw new IllegalArgumentException("Value [" + bigIntegerValue + "] is out of range for unsigned long");
+            }
+            return bigIntegerValue.longValue();
+        } else { // for all other value types including numbers with decimal parts
+            throw new IllegalArgumentException("For input string: [" + parser.getValueAsString() + "].");
+        }
+    }
```

Comment:
do we really need to have support for unsigned longs at such a low level, or could we implement parsing of unsigned longs in the field mapper directly?